src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -1520,11 +1520,12 @@ abstract class RefChecks extends InfoTransform with scala.reflect.internal.trans
       }
 
       val doTransform =
+        !tree.tpe.isInstanceOf[PolyType] && // SI-9546 If we've bailed out of the transform for a TypeApply, don't try again for the enclosed Select
```

Comment:
This one actually makes more sense as the primary check. Otherwise, in the case of a public constructor, we are relying on the details of the transform (e.g. that parents are transformed before children).